Core/include/Acts/Propagator/Navigator.hpp

```diff
@@ -393,904 +336,382 @@ class Navigator {
       // Set navigation break and release the navigation step size
       state.navigation.navigationBreak = true;
       stepper.releaseStepSize(state.stepping, ConstrainedStep::actor);
     }
 
-    // Navigator target always resets the current surface
+    // Navigator preStep always resets the current surface
     state.navigation.currentSurface = nullptr;
   }
 
-  /// @brief Navigator post step call, will be called in two modes
-  ///
-  /// (a) It initializes the Navigation stream if start volume is
-  ///     not yet defined:
-  ///  - initialize the volume
-  ///  - establish the start layer and start volume
-  ///  - set the current surface to the start surface
+  /// @brief Navigator post step call
   ///
-  /// (b) It establishes the currentSurface status during
-  ///     the propagation flow, currentSurface can be
-  ///  - surfaces still to be handled within a layer
-  ///  - layers still to be handled within a volume
-  ///  - boundaries still to be handled to exit a volume
+  /// This determines if we hit the next navigation candidate and deals with it
+  /// accordingly. It sets the current surface, enters layers and changes
+  /// volumes.
   ///
   /// @tparam propagator_state_t is the type of Propagatgor state
   /// @tparam stepper_t is the used type of the Stepper by the Propagator
   ///
   /// @param [in,out] state is the mutable propagator state object
   /// @param [in] stepper Stepper in use
+  ///
+  /// @return Boolean to indicate if we continue with the actors and
+  ///         aborters or if we should target again.
   template <typename propagator_state_t, typename stepper_t>
-  void postStep(propagator_state_t& state, const stepper_t& stepper) const {
+  bool postStep(propagator_state_t& state, const stepper_t& stepper) const {
+    ACTS_VERBOSE(volInfo(state) << "Entering navigator::postStep.");
+
     // Check if the navigator is inactive
-    if (inactive(state, stepper)) {
-      return;
+    if (inactive(state)) {
+      return true;
     }
 
-    // Set the navigation stage
-    state.navigation.navigationStage = Stage::undefined;
+    assert(state.navigation.currentSurface == nullptr &&
+           "Current surface must be reset.");
 
-    // Call the navigation helper prior to actual navigation
-    ACTS_VERBOSE(volInfo(state) << "Entering navigator::postStep.");
+    const auto& candidate = state.navigation.candidate();
+    const auto& intersection = candidate.intersection;
+    const Surface& surface = *intersection.representation();
 
-    // Navigator post step always starts without current surface
-    state.navigation.currentSurface = nullptr;
+    // Determine the surface status without boundary check.
+    // TODO This will also update the step length which is unnecessary
+    Intersection3D::Status surfaceStatus = stepper.updateSurfaceStatus(
+        state.stepping, surface, intersection.index(), state.options.direction,
+        BoundaryCheck(false), state.options.surfaceTolerance, logger());
```

Comment:
ah I remember now @paulgessinger - in principle we can be out of bounds but on surface after the step while we thought we would be in bounds before the step

if this is a sensitive it does not really matter and we will skip it in the next `preStep` call. for layers and boundaries this would cause a renavigation and we might be out of the current volume in an unfortunate edge case. in such a case it would be nice to see a log that we actually hit the surface but out of bounds

I guess this is more of a debugging case then but if the performance hit is not significant I would like to keep it